### PR TITLE
Fix #6870 Encode control characters

### LINKED ISSUE #6870: Encode control characters in URIUtil.encodePath
**Jetty version(s)**
9.4.43 and 10.0.1

**Description**
Runing Jetty with DefaultServlet, `http://localhost/%0A` returns a 404 - as expected.

However, when using BalancerServlet, the same request URL causes a 500 error. Same goes for anything from %01 to %1F (i.e. encoded control characters).

By simplifying to a minimal example, it turns out the issue does not occur without the rewrite module enabled.
More specifically, it still doesn't occur if only a RewriteHandler has been added, but adding a VirtualHostRuleContainer triggers it, whether or not there are rules within that container.

Based on the exception, _something_ is decoding the request URL before it is being passed to a URI.create() call, but given that `http://localhost/%20` does NOT trigger the error, it's also being _partially_ re-encoded (otherwise URI.create would complain about the space character, which it doesn't).

So, looking at the code, the bug will most likely be in `URIUtil.encodePath` - perhaps the `(c < 0)` should be a `(c < 20)` or `(c < 32 )` ? (either way, looks to me like that function could be cleaned up).

There may also be a secondary issue - is `RuleContainer.apply` being called when it doesn't need to be?
(if the container is either empty, or the virtualhosts don't match the current request, it should be a no-op?)

**How to reproduce?**
Bash script demonstrating the issue:

	JETTY_HOME=/path/to/jetty-home-9.4.43.v20210629
	JETTY_BASE=test-badurl

	mkdir -p "$JETTY_BASE"/{webapps,modules,etc}
	sed 's!etc/jetty-proxy.xml!#&!' "$JETTY_HOME/modules/proxy.mod" > "$JETTY_BASE/modules/rproxy.mod"

	echo '<?xml version="1.0" encoding="UTF-8"?>
	<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure.dtd">
	<Configure class="org.eclipse.jetty.servlet.ServletContextHandler">
	 <Call name="addServlet">
	  <Arg>org.eclipse.jetty.proxy.BalancerServlet</Arg>
	  <Arg>/*</Arg>
	  <Call name="setInitParameter">
	   <Arg>balancerMember.localhost_8123.proxyTo</Arg>
	   <Arg>http://localhost:8123/</Arg>
	  </Call>
	 </Call>
	</Configure>' > "$JETTY_BASE/webapps/ROOT.xml"

	echo '<?xml version="1.0"?>
	<!DOCTYPE Configure PUBLIC "-//Jetty//Configure//EN" "http://www.eclipse.org/jetty/configure_9_0.dtd">
	<Configure id="Server" class="org.eclipse.jetty.server.Server">
	 <Call name="insertHandler">
	  <Arg>
	   <New class="org.eclipse.jetty.rewrite.handler.RewriteHandler">
	    <Set name="originalPathAttribute">requestedPath</Set>
	    <Set name="rewriteRequestURI">true</Set>
	    <Set name="rules">
	     <Array type="org.eclipse.jetty.rewrite.handler.Rule">
	      <Item>
	       <New class="org.eclipse.jetty.rewrite.handler.VirtualHostRuleContainer">
	       </New>
	      </Item>
	     </Array>
	    </Set>
	   </New>
	  </Arg>
	 </Call>
	</Configure>' > "$JETTY_BASE/etc/jetty-rewrite.xml"

	cd "$JETTY_BASE"
	java -jar $JETTY_HOME/start.jar --add-to-start=http,deploy,rproxy,rewrite

	java -jar $JETTY_HOME/start.jar &

	curl -isS 'http://localhost:8080/%0A'

Related exception:

	2021-09-19 21:31:30.505:WARN:oejs.HttpChannel:qtp5592464-19: /|
	java.lang.IllegalArgumentException: Illegal character in path at index 24: http://localhost:8123///|
	        at java.net.URI.create(Unknown Source)
	        at org.eclipse.jetty.proxy.BalancerServlet.rewriteTarget(BalancerServlet.java:142)
	        at org.eclipse.jetty.proxy.ProxyServlet.service(ProxyServlet.java:61)
	        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
	        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
	        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	        at org.eclipse.jetty.server.Server.handle(Server.java:516)
	        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
	        at org.eclipse.jetty.server.HttpChannel$$Lambda$87/1776338520.dispatch(Unknown Source)
	        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
	        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
	        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	        at java.lang.Thread.run(Unknown Source)
	Caused by:
	java.net.URISyntaxException: Illegal character in path at index 24: http://localhost:8123///|
	        at java.net.URI$Parser.fail(Unknown Source)
	        at java.net.URI$Parser.checkChars(Unknown Source)
	        at java.net.URI$Parser.parseHierarchical(Unknown Source)
	        at java.net.URI$Parser.parse(Unknown Source)
	        at java.net.URI.<init>(Unknown Source)
	        at java.net.URI.create(Unknown Source)
	        at org.eclipse.jetty.proxy.BalancerServlet.rewriteTarget(BalancerServlet.java:142)
	        at org.eclipse.jetty.proxy.ProxyServlet.service(ProxyServlet.java:61)
	        at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
	        at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
	        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:191)
	        at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	        at org.eclipse.jetty.rewrite.handler.RewriteHandler.handle(RewriteHandler.java:322)
	        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	        at org.eclipse.jetty.server.Server.handle(Server.java:516)
	        at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:388)
	        at org.eclipse.jetty.server.HttpChannel$$Lambda$87/1776338520.dispatch(Unknown Source)
	        at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:633)
	        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:380)
	        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	        at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	        at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	        at java.lang.Thread.run(Unknown Source)

/end

### PR DESCRIPTION
**Pull request by gregw**
Reproduce #6870 Rewritten Balancer

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
I've pushed a reproduction in a unit test for the BalancerServlet, although ultimately I think this should be a test/fix in jetty-http and/or jetty-rewrite

File: jetty-proxy/src/test/java/org/eclipse/jetty/proxy/BalancerServletTest.java [hunk 1]
```diff
@@ -111,6 +116,15 @@ private int getServerPort(Server server)
         return server.getURI().getPort();
     }
 
+    protected ContentResponse getBalancedResponse(String path) throws Exception
+    {
+        ContentResponse response = client.newRequest("localhost", getServerPort(balancer))
+            .path(CONTEXT_PATH + SERVLET_PATH + path)
+            .timeout(5, TimeUnit.SECONDS)
+            .send();
+        return response;
```
Comment (sbordet) [inline jetty-proxy/src/test/java/org/eclipse/jetty/proxy/BalancerServletTest.java]: Inline `response` variable.

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 2]
```diff
@@ -120,7 +120,7 @@ private static StringBuilder encodePath(StringBuilder buf, String path, int offs
                         buf = new StringBuilder(path.length() * 2);
                         break loop;
                     default:
-                        if (c > 127)
+                        if (c < 20 || c > 127)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Decimal `20` is not good I guess? If you means the space it's decimal `32`, or hex `0x20`.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-proxy/src/test/java/org/eclipse/jetty/proxy/BalancerServletTest.java (hunk 1 above)
Comment (gregw) [reply jetty-proxy/src/test/java/org/eclipse/jetty/proxy/BalancerServletTest.java]: done

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 2 above)
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: ooops.  Added a better test that checks the whole range of characters.  Also encode 127, which is DEL!

**Review by joakime — APPROVED** [review]

**Review by sbordet — APPROVED** [review]

**Pull request merged**